### PR TITLE
[threadpool-io] Extra asserts to trackdown `fd >=0` assertion

### PR DESCRIPTION
**Pull request by jaykrell**
Diagnosability for #14848.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by lambdageek** [comment]
We're probably not going to merge this (or at least not right away), but it may be useful to cherrypick this into a custom Xamarin.Android build to track down the underlying issue.

**Comment by jaykrell** [comment]
It is asserts, not logging. Not to nitpick title, but would want to decide which is prefered.

**Comment by jaykrell** [comment]
An in-memory log, ringbuffer, of the last n items, not a terrible idea.
And/or change union to struct, wasteful but probably not horrible.

File: mono/metadata/threadpool-io.c [hunk 1]
```diff
@@ -651,20 +664,18 @@ mono_threadpool_io_remove_socket (int fd)
...
-	update = update_get_new ();
-	update->type = UPDATE_REMOVE_SOCKET;
-	update->data.add.fd = fd;
-	mono_memory_barrier (); /* Ensure this is safely published before we wake up the selector */
+	if (io_selector_running) {
 
-	selector_thread_wakeup ();
+		update = update_get_new (UPDATE_REMOVE_SOCKET, __func__);
+		g_assertf (fd >= 0, "%d", fd);
+		update->data.remove_socket.fd = fd;
```
Comment (lambdageek) [inline mono/metadata/threadpool-io.c]: Kind of lost in here, but we should take this change no matter what.
`update->data.add.fd = fd` to `update->data.remove_socket.fd = fd`

**Comment by CoffeeFlux** [comment]
We tracked down that particular bug, so I don't think there's a need to merge this unless the refactoring changes are otherwise important.

**Comment by jaykrell** [comment]
Agreed.

**Pull request closed without merging**